### PR TITLE
Adding fixed atom constraints for NVT simulations

### PR DESCRIPTION
**Pull request by lucasdekam**
Hello,

First of all, really nice job on this lightweight package and using JAX-MD is also a cool idea. 

I'm interested in doing simulations with a slab geometry where the bottom atoms of the slabs are fixed. There's no native functionality for this in JAX-MD but I tried to implement it with a wrapper.

Initially I just wrapped the `shift_fn` from
```
self._displacement_fn, self._shift_fn = jax_md.space.periodic_general(self._initial_box, fractional_coordinates="NPT" in self._routine)
```
with a masked shift function (see [this thread](https://github.com/jax-md/jax-md/issues/115)), but this doesn't work for NVT simulations. In fact, when I tried it, all atoms in the simulation slowly froze. The JAX-MD issue thread also suggested that the reason is that degrees of freedom aren't properly accounted for. With the help of ChatGPT I ended up with a wrapper function that corrects the DoF and sets the momentum of constrained atoms to zero after the NVT function update. 

This is more of a draft PR - I'm a beginner when it comes to MD simulation code so I'd greatly appreciate any help. In particular,

* any idea whether this approach is physically sound?
* would it be a good idea to apply the wrapper to all of the NVE, NVT and NPT simulation routines?
* any things I should test in particular to verify the code? 
* should the wrapped init and apply functions be jitted? (in my test, it didn't seem to affect speed, but maybe it does for bigger systems)

### CODE REVIEW

File: deepmd_jax/md.py [hunk 1]
```diff
@@ -319,8 +350,24 @@ def __init__(self,
...
+
+        # Create mask for constraining atoms
+        mobile = np.ones(self._natoms)
+        if fixed_indices is not None:
+            mobile[fixed_indices] = 0
+            self._mobile = jnp.array(mobile, dtype=bool)
+        else: 
+            self._mobile = None
+
+        def _shift_fn_wrapper(x, dx, **kwargs):
```
Comment (lucasdekam) [inline deepmd_jax/md.py:365]: Currently, the wrapper is applied for all routines, maybe not desired

Comment (lucasdekam) [reply deepmd_jax/md.py:365]: Thanks for the feedback! I added the `shift_fn` wrapper because atoms were still moving a tiny bit when only adapting the NVT wrapper. As-is this works for me, but I have only tested NVT so far.

I can polish up the implementation a bit, let me know if you have any suggestions. I can try to make it general for NVE,NVT and NPT, but don't have any data that allows me to test the NPT dynamics.

Might take me a few weeks to get to it properly, got some stuff to do and vacation coming up :)

Comment (lucasdekam) [reply deepmd_jax/md.py]: Just to update, after running quite a few trajectories on an Au/water system with the bottom 2 Au layers fixed, the invariants seem pretty constant. I attached a typical logfile. The invariant does change a tiny bit over time. 

[deepmd_jax.txt](https://github.com/user-attachments/files/23795577/deepmd_jax.txt)

**Comment by SparkyTruck** [comment]
I have update the code, made some reorganization, and push a commit in a new [branch](https://github.com/SparkyTruck/deepmd-jax/tree/pr-16).
I added additional hard constraints on the energy function to zero out forces for fixed atoms. This might have had something to do inside the NVT thermostat -- now from my side things appear to be working properly.
If it works well for you as well, we can merge this PR.
